packages/react/src/select/popup/SelectPopup.tsx

```diff
@@ -216,131 +217,165 @@ export const SelectPopup = React.forwardRef(function SelectPopup(
-      const maxHeight = viewportHeight - marginTop - marginBottom;
-      const scrollTop = idealHeight - height;
+      // Ensure we remove any transforms that can affect the location of the popup
+      // and therefore the calculations.
+      const originalTransform = popupElement.style.transform;
+      popupElement.style.transform = 'none';
+      popupElement.style.removeProperty('--transform-origin');
+
+      try {
+        const positionerStyles = getComputedStyle(positionerElement);
+        const popupStyles = getComputedStyle(popupElement);
+
+        const doc = ownerDocument(triggerElement);
+        const win = ownerWindow(positionerElement);
+        const triggerRect = triggerElement.getBoundingClientRect();
+        const positionerRect = positionerElement.getBoundingClientRect();
+        const triggerX = triggerRect.left;
+        const triggerHeight = triggerRect.height;
+        const scroller = listElement || popupElement;
+        const scrollHeight = scroller.scrollHeight;
+
+        const borderBottom = parseFloat(popupStyles.borderBottomWidth);
+        const marginTop = parseFloat(positionerStyles.marginTop) || 10;
+        const marginBottom = parseFloat(positionerStyles.marginBottom) || 10;
+        const minHeight = parseFloat(positionerStyles.minHeight) || 100;
+
+        const paddingLeft = 5;
+        const paddingRight = 5;
+        const triggerCollisionThreshold = 20;
+
+        const viewportHeight = doc.documentElement.clientHeight - marginTop - marginBottom;
+        const viewportWidth = doc.documentElement.clientWidth;
+        const availableSpaceBeneathTrigger = viewportHeight - triggerRect.bottom + triggerHeight;
+
+        const textElement = selectedItemTextRef.current;
+        const valueElement = valueRef.current;
+        let offsetX = 0;
+        let offsetY = 0;
+
+        if (textElement && valueElement) {
+          const valueRect = valueElement.getBoundingClientRect();
+          const textRect = textElement.getBoundingClientRect();
+
+          const valueLeftFromTriggerLeft = valueRect.left - triggerX;
+          const textLeftFromPositionerLeft = textRect.left - positionerRect.left;
+          const valueCenterFromPositionerTop =
+            valueRect.top - triggerRect.top + valueRect.height / 2;
+          const textCenterFromTriggerTop = textRect.top - positionerRect.top + textRect.height / 2;
+
+          offsetX = valueLeftFromTriggerLeft - textLeftFromPositionerLeft;
+          offsetY = textCenterFromTriggerTop - valueCenterFromPositionerTop;
+
+          const popupTop = positionerRect.top;
+          const popupHeight = positionerRect.height;
+          const textCenterY = textRect.top + textRect.height / 2;
+
+          const transformOriginY =
+            popupHeight > 0 ? ((textCenterY - popupTop) / popupHeight) * 100 : 50;
+
+          const clampedY = clamp(transformOriginY, 0, 100);
+
+          popupElement.style.setProperty('--transform-origin', `50% ${clampedY}%`);
```

Comment:
Since the addition of `try {} finally {}` breaks the diff:

This part is new: it maintains a correct transform origin based on the location of the item.